### PR TITLE
add http request timeout for OpenStack cloud provider

### LINKED ISSUE #53191: All endpoint controller workers stuck at http call because of missing timeout setting
**What happened**:
We use OpenStack as our cloud provider. We meet an issue that some controllers stuck because all workers routine stuck at http call to OpenStack services because of rabbitmq issues.

Stuck routine call stack:
```
goroutine 65108 [select, 145 minutes]:
net/http.(*persistConn).roundTrip(0xc42f6fb700, 0xc422c5acc0, 0x0, 0x0, 0x0)
    /usr/local/go/src/net/http/transport.go:1840 +0x93b
net/http.(*Transport).RoundTrip(0xc4200f0000, 0xc42547f950, 0xc4200f0000, 0x0, 0xc400000000)
    /usr/local/go/src/net/http/transport.go:380 +0x4fa
net/http.send(0xc42547f950, 0x48e4b80, 0xc4200f0000, 0x0, 0x0, 0x0, 0x8, 0xc42b3cea20, 0x40fb78)
    /usr/local/go/src/net/http/client.go:256 +0x15f
net/http.(*Client).send(0xc4211c8428, 0xc42547f950, 0x0, 0x0, 0x0, 0xc42b3cea20, 0x0, 0x1)
    /usr/local/go/src/net/http/client.go:146 +0x102
net/http.(*Client).doFollowingRedirects(0xc4211c8428, 0xc42547f950, 0x33407f0, 0x4, 0x1, 0xc42ae6a3c8)
    /usr/local/go/src/net/http/client.go:528 +0x5e5
net/http.(*Client).Do(0xc4211c8428, 0xc42547f950, 0x1c, 0x3177153, 0x11)
    /usr/local/go/src/net/http/client.go:187 +0x11d
k8s.io/kubernetes/vendor/github.com/gophercloud/gophercloud.(*ProviderClient).Request(0xc4211c83f0, 0x3159d93, 0x4, 0xc428126c00, 0x37, 0xc430a750a8, 0x0, 0x0, 0x0)
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/gophercloud/gophercloud/provider_client.go:169 +0x3b5
k8s.io/kubernetes/vendor/github.com/gophercloud/gophercloud.(*ServiceClient).Post(0xc426444e40, 0xc428126c00, 0x37, 0x2bff4a0, 0xc4289f1500, 0x298cde0, 0xc4289f1470, 0xc430a750a8, 0x40c99f, 0x2f21380, ...)
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/gophercloud/gophercloud/service_client.go:78 +0x1b1
k8s.io/kubernetes/vendor/github.com/gophercloud/gophercloud/openstack/networking/v2/extensions/lbaas/members.Create(0xc426444e40, 0x48e76c0, 0xc428126b80, 0x0, 0x0, 0x0, 0x0, 0x0)
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/gophercloud/gophercloud/openstack/networking/v2/extensions/lbaas/members/requests.go:74 +0x19a
```

Checked the openstack cloud provider that there is no HttpClient.Timeout set which means no timeout during http call. I think we should set at least the default http timeout value.
https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/providers/openstack/openstack.go#L264-L291

**Environment**:
- Kubernetes version (use `kubectl version`):
v1.6.3
- Cloud provider or hardware configuration**:
OpenStack

### PR DESCRIPTION
**Pull request by dixudx**
**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #53191

**Special notes for your reviewer**:
/assign @NickrenREN @dims @FengyunPan 

**Release note**:

```release-note
None
```

### CODE REVIEW

File: pkg/cloudprovider/providers/openstack/openstack.go [hunk 1]
```diff
@@ -290,6 +291,7 @@ func newOpenStack(cfg Config) (*OpenStack, error) {
 		return nil, err
 	}
 
+	provider.HTTPClient.Timeout = cfg.Metadata.RequestTimeout
```
Comment (yastij) [inline pkg/cloudprovider/providers/openstack/openstack.go]: @dixudx - we should set also default timeout value if none provided.

Comment (xinxiaogang) [reply pkg/cloudprovider/providers/openstack/openstack.go]: In gophercloud, if no value is set, the default value is zero which means no timeout.

**Comment by dixudx** [comment]
@yastij @xinxiaogang Updated. PTAL.

**Comment by NickrenREN** [comment]
Thanks for fixing this @dixudx . But i do not think it is a good option to add another param to the config file. As we said before it would become more and more complex for users to configure.
Maybe a default const var is enough ? WDYT

**Comment by dixudx** [comment]
> As we said before it would become more and more complex for users to configure.
Maybe a default const var is enough ?

@NickrenREN Make sense. No need to expose extra params unless it is meaningful and important. 

About the default value, I think setting `60` as default value is enough, WDYT?

**Comment by NickrenREN** [comment]
@dixudx  I am not so sure what the `defaultTimeOut` should be. Need feedback from more people.
Thinking more, do we need backoff policy for some requests ? What if the request needs more time to reach than the `Need more discussion` ? Just return error directly ?

**Comment by xinxiaogang** [comment]
Do we have backoff logic as a general component in controller-manager ? I think not only endpoint controller need this to call downstream APIs.

**Comment by yastij** [comment]
from what i recall `60` is the one set on aws (ping @justinsb)

**Comment by dixudx** [comment]
> or the request failed just now but can succeed next time

@NickrenREN I don't think k8s should be responsible for this. What k8s can do is to add some retrys.

> Just return error directly when the request failed once ?

Seems we're making it more complicated. All the actual requests are performed by `Gophercloud`, it will be error-prone to handle this here.

> What if the request needs more time to reach than the `defaultTimeOut`

Per above discussions, I think we'd better leave an option for the users to customize the timeOut by their own.

**Comment by yastij** [comment]
@dixudx - for those kind of things, I prefer having the same behaviour as much as possible

**Comment by dims** [comment]
/ok-to-test

Is it in seconds?

**Comment by dixudx** [comment]
> Is it in seconds?

@dims Yes, in seconds.

**Comment by yastij** [comment]
@NickrenREN @dims - this LGTY ?

File: pkg/cloudprovider/providers/openstack/openstack.go [hunk 2]
```diff
@@ -52,6 +52,7 @@ import (
 const (
 	ProviderName     = "openstack"
 	AvailabilityZone = "availability_zone"
+	defaultTimeOut   = 60
```
Comment (dims) [inline pkg/cloudprovider/providers/openstack/openstack.go]: If this is seconds, it should be `time.Duration(60*time.Second)` Right?

https://play.golang.org/p/gmIFXzUgQF

Comment (yastij) [reply pkg/cloudprovider/providers/openstack/openstack.go]: yup indeed

**Comment by dixudx** [comment]
@dims @yastij Updated. PTAL.

**Comment by dixudx** [comment]
ping @NickrenREN Needs your `lgtm`. Thanks.

**Pull request merged**